cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -228,51 +233,87 @@ private Table getMetaTable(DatasetContext context) throws IOException, DatasetMa
     try {
       return Transactions.execute(transactional, new TxCallable<List<ArtifactDetail>>() {
         @Override
         public List<ArtifactDetail> call(DatasetContext context) throws Exception {
-          return getArtifacts(getMetaTable(context), range);
+          return getArtifacts(getMetaTable(context), range, limit, order);
         }
       });
     } catch (TransactionFailureException e) {
       throw Transactions.propagate(e);
     }
   }
 
-  private List<ArtifactDetail> getArtifacts(Table metaTable, ArtifactRange range) {
+  private List<ArtifactDetail> getArtifacts(
+    Table metaTable, ArtifactRange range, int limit,
+    @Nullable SortInfo.SortOrder order) throws IOException, ArtifactRangeNotFoundException, ArtifactNotFoundException {
     List<ArtifactDetail> artifacts = Lists.newArrayList();
     ArtifactKey artifactKey = new ArtifactKey(range.getNamespace(), range.getName());
 
     Row row = metaTable.get(artifactKey.getRowKey());
+    if (order == null) {
+      addArtifactsToList(artifacts, row, limit, range);
+    } else {
+      addAndSortArtifacts(artifacts, row, limit, order, range);
+    }
+    return Collections.unmodifiableList(artifacts);
+  }
+
+  private void addAndSortArtifacts(List<ArtifactDetail> artifacts, Row row, int limit,
+                                   @Nullable SortInfo.SortOrder order, @Nullable ArtifactRange range) {
+    ArtifactKey artifactKey = ArtifactKey.parse(row.getRow());
+    SortedMap<ArtifactId, ArtifactData> sortedArtifacts = order == SortInfo.SortOrder.DESC ?
+      new TreeMap<ArtifactId, ArtifactData>(Collections.<ArtifactId>reverseOrder()) :
+      new TreeMap<ArtifactId, ArtifactData>();
+
     for (Map.Entry<byte[], byte[]> columnEntry : row.getColumns().entrySet()) {
       String version = Bytes.toString(columnEntry.getKey());
-      if (range.versionIsInRange(new ArtifactVersion(version))) {
-        ArtifactData data = GSON.fromJson(Bytes.toString(columnEntry.getValue()), ArtifactData.class);
-        Id.Artifact artifactId = Id.Artifact.from(artifactKey.namespace.toId(), artifactKey.name, version);
-        artifacts.add(new ArtifactDetail(new ArtifactDescriptor(
-          artifactId.toArtifactId(),
-          Locations.getLocationFromAbsolutePath(locationFactory, data.getLocationPath())), data.meta));
+      if (range != null && !range.versionIsInRange(new ArtifactVersion(version))) {
+        continue;
       }
+
+      ArtifactData data = GSON.fromJson(Bytes.toString(columnEntry.getValue()), ArtifactData.class);
+      sortedArtifacts.put(
+          new ArtifactId(artifactKey.name, new ArtifactVersion(version),
+                         artifactKey.namespace.equals(NamespaceId.SYSTEM) ? ArtifactScope.SYSTEM : ArtifactScope.USER),
+          data);
+    }
+    for (Map.Entry<ArtifactId, ArtifactData> entry : sortedArtifacts.entrySet()) {
+      if (limit != -1 && artifacts.size() == limit) {
```

Comment:
try to avoid special values if possible. Way back in the http handler, we should be translating -1 to max integer value.